src/test/on-error.test.ts

```diff
@@ -0,0 +1,284 @@
+import { Server, onError, CallType } from '..'
+import {
+  GreetingService,
+  GreetingClient,
+} from '../../dist/test/api/v1/hello_grpc_pb'
+import {
+  ServerCredentials,
+  ChannelCredentials,
+  Metadata,
+  StatusObject,
+} from '@grpc/grpc-js'
+import { Hello } from '../../dist/test/api/v1/hello_pb'
+
+const ADDR = '0.0.0.0:3000'
+describe('Error handling', () => {
+  const server = new Server()
+  let lastError: any = null
+  afterAll(() => server.stop())
+  test('Setup', async () => {
+    server.addService(GreetingService, {
+      unary: ctx => {
+        throw new Error('unary-error')
+      },
+      serverStream: ctx => {
+        if (ctx.metadata.getMap().type === 'sync') {
+          throw new Error('serverStream-sync-error')
+        } else {
+          ctx.emit('error', new Error('serverStream-stream-error'))
+        }
+      },
+      clientStream: ctx => {
+        if (ctx.metadata.getMap().type === 'sync') {
+          throw new Error('clientStream-sync-error')
+        } else {
+          ctx.emit('error', new Error('clientStream-stream-error'))
+        }
+      },
+      bidi: ctx => {
+        if (ctx.metadata.getMap().type === 'sync') {
+          throw new Error('bidi-sync-error')
+        } else {
+          ctx.emit('error', new Error('bidi-stream-error'))
+        }
+      },
+    })
+    server.use(
+      onError((e, ctx) => {
+        ctx.initialMetadata.set('onerror', `${ctx.type}-onerror`)
+        ctx.trailingMetadata.set('onerror', `${ctx.type}-onerror`)
+        if (ctx.metadata.getMap().catch) {
+          lastError = e
+          if (
+            ctx.type === CallType.SERVER_STREAM ||
+            ctx.type === CallType.BIDI
+          ) {
+            // sync error not re-thrown on stream response, should end
+            ctx.end()
+          }
+        } else {
+          // Handler can be sync or async
+          if (Math.random() > 0.5) {
+            throw e
+          }
+          return Promise.reject(e)
+        }
+      })
+    )
+    server.use((ctx, next) => {
+      ctx.initialMetadata.set('initial', ctx.type)
+      ctx.trailingMetadata.set('trailing', `${ctx.type}-trailing`)
+    })
+    await server.start(ADDR, ServerCredentials.createInsecure())
+  })
+  describe('Unary', () => {
+    const client = new GreetingClient(ADDR, ChannelCredentials.createInsecure())
+
+    const clientMeta = new Metadata()
+    clientMeta.set('catch', 'true')
+    test('Caught error does not reach client', async () => {
+      await new Promise<Hello>((resolve, reject) => {
```

Comment:
Why not use `promisify(client.unary)`? Would make the tests easier to read. It repeats here alot!

Speaking easy to read - maybe you should consider using a helper for streams `(stream) => Promise<{ data, error /* other events. */ }>`. A small helper, unusable in production ofc, it breaks the principle of streaming - but it will make the repeated code shorter .]

Reply:
I agree the tests would deserve some love. I'd rather solve this later and refactor it with existing integration tests. I think it's a good opportunity to develop and async client with more enjoyable interface in the near future.